### PR TITLE
Test-SqlMigrationConstraint, update for new features of 2016 SP1

### PR DESCRIPTION
**Pull request by ClaudioESSilva**
Take into account the great changes on SQL Server 2016 SP1 which have all features available  on 'all' editions

### CODE REVIEW

**Comment by ClaudioESSilva** [comment]
Based on @ctrlbold lab.

If you run the following code (using SQLCMD Mode - Query Menu -> SQLCMD Mode):
```
:connect sql2016
SELECT @@SERVERNAME AS ServerName, SERVERPROPERTY('ProductLevel') AS ProductLevel, SERVERPROPERTY('Edition') AS Edition, SERVERPROPERTY('ProductVersion') AS ProductVersion
GO

:connect sql2016a
SELECT @@SERVERNAME AS ServerName, SERVERPROPERTY('ProductLevel') AS ProductLevel, SERVERPROPERTY('Edition') AS Edition, SERVERPROPERTY('ProductVersion') AS ProductVersion
GO

:connect sql2016\STANDARDRTM
SELECT @@SERVERNAME AS ServerName, SERVERPROPERTY('ProductLevel') AS ProductLevel, SERVERPROPERTY('Edition') AS Edition, SERVERPROPERTY('ProductVersion') AS ProductVersion
GO

:connect sql2016\sqlexpress
SELECT @@SERVERNAME AS ServerName, SERVERPROPERTY('ProductLevel') AS ProductLevel, SERVERPROPERTY('Edition') AS Edition, SERVERPROPERTY('ProductVersion') AS ProductVersion
GO

:connect sql2012
SELECT @@SERVERNAME AS ServerName, SERVERPROPERTY('ProductLevel') AS ProductLevel, SERVERPROPERTY('Edition') AS Edition, SERVERPROPERTY('ProductVersion') AS ProductVersion
GO
```

Returns:

| ServerName          | ProductLevel | Edition                     | ProductVersion |
|---------------------|--------------|-----------------------------|----------------|
| SQL2016             | SP1          | Standard Edition (64-bit)   | 13.0.4001.0    |
| SQL2016A            | RTM          | Enterprise Edition (64-bit) | 13.0.1601.5    |
| SQL2016\STANDARDRTM | RTM          | Standard Edition (64-bit)   | 13.0.1601.5    |
| SQL2016\SQLEXPRESS  | SP1          | Express Edition (64-bit)    | 13.0.4001.0    |
| SQL2012             | SP3          | Enterprise Edition (64-bit) | 11.0.6020.0    |


Use the following parameters to test the improvements:

- [ ] `Test-SqlMigrationConstraint -Source SQL2016 -Destination SQL2016A` -> Expected result: For every database should say "Database can be migrated"
- [ ] `Test-SqlMigrationConstraint -Source SQL2016A -Destination SQL2016` -> Expected result: For every database should say "Database can be migrated"
- [ ] `Test-SqlMigrationConstraint -Source SQL2016A -Destination SQL2016\STANDARDRTM` -> Expected result: For every database that have features should say "Database cannot be migrated. There are features in use that are not available on destination instance edition." otherwise "Database can be migrated"
- [ ] `Test-SqlMigrationConstraint -Source SQL2016 -Destination SQL2016\STANDARDRTM` -> Expected result: For every database that have features should say "Database cannot be migrated. There are features in use that are not available on destination instance edition." otherwise "Database can be migrated"
- [ ] `Test-SqlMigrationConstraint -Source SQL2016A -Destination SQL2016\SQLEXPRESS` -> Expected result: For every database that have 'ChangeCapture' feature should say "Database cannot be migrated. Destination server is running Express Edition which does not support 'ChangeCapture' feature." otherwise "Database can be migrated"
- [ ] `Test-SqlMigrationConstraint -Source SQL2012 -Destination SQL2016\SQLEXPRESS -Databases DBWithCompression` -> Expected result: "Database can be migrated"

**Please feel free to try other combinations and ask at will if you have any questions.**

If you want to create some databases with features you can use the following scripts:

Database with Compression
```
CREATE DATABASE DBWithCompression
GO

USE DBWithCompression
GO

CREATE TABLE TableCompressed
(
	C1 INT
)

CREATE INDEX CI__TableCompressed ON TableCompressed
(
	C1
) WITH (DATA_COMPRESSION = PAGE)

-- Show features is in use
SELECT * FROM sys.dm_db_persisted_sku_features
GO
```


Database with Change Data Capture (CDC)
```
CREATE DATABASE DBWithCDC
GO

USE DBWithCDC
GO  

CREATE TABLE TableCDC
(
	C1 INT
)
GO
 
EXEC sys.sp_cdc_enable_db  
GO 

EXECUTE sys.sp_cdc_enable_table  
    @source_schema = N'dbo'  
  , @source_name = N'TableCDC'  
  , @role_name = N'cdc_Admin';  
GO   

-- Show features is in use
SELECT * FROM sys.dm_db_persisted_sku_features
GO
```

Database with Clustered Column Store Index (CCSI)
```
-- Only for SQL Server 2016
CREATE DATABASE DBWithCSI
GO

USE DBWithCSI
GO

CREATE TABLE TableCSI
(
	C1 INT
)

CREATE CLUSTERED COLUMNSTORE INDEX CCI__TableCSI ON TableCSI;  
GO  


-- Verify that feature is in use
SELECT * FROM sys.dm_db_persisted_sku_features
GO
```

**Comment by ctrlbold** [comment]
Holy shit! You get like 8 🥇🥇🥇🥇🥇🥇🥇🥇for the level of detail in this PR! Thank you @ClaudioESSilva. Looks like I'll have time to work on this tonight!

File: functions/Test-SqlMigrationConstraint.ps1 [hunk 1]
```diff
@@ -180,50 +187,116 @@ Only db1 database will be verified for features in use that can't be supported o
                         $dbstatus = $db.Status.ToString()
                         $dbName = $db.Name
                     }
-                    Write-Host "`r`nChecking database: '$dbName'"
```
Comment (wsmelton) [inline functions/Test-SqlMigrationConstraint.ps1]: Should also say `Write-Output` and not `Write-Host`

**Comment by ctrlbold** [comment]
This was a really gorgeous PR. Thanks so much for the tests! Unfortunately, there's a stupid conflict that I can't resolve. Please resubmit.

File: functions/Test-SqlMigrationConstraint.ps1 (hunk 1 above)
Comment (ctrlbold) [inline functions/Test-SqlMigrationConstraint.ps1]: Ah ha! true D:

**Pull request closed without merging**